nova/cmd/manage.py

```diff
@@ -2453,14 +2454,174 @@ def sync_aggregates(self, verbose=False):
+            try:
+                im = objects.InstanceMapping.get_by_instance_uuid(
+                    ctxt, instance_uuid)
+                cells = objects.CellMappingList(objects=[im.cell_mapping])
+            except exception.InstanceMappingNotFound:
+                print('Unable to find cell for instance %s, is it mapped? Try '
+                      'running "nova-manage cell_v2 verify_instance" or '
+                      '"nova-manage cell_v2 map_instances".' %
+                      instance_uuid)
+                return 127
+        else:
+            cells = objects.CellMappingList.get_all(ctxt)
+            if not cells:
+                output(_('No cells to process.'))
+                return 4
+
+        for cell in cells:
+            if cell.uuid == objects.CellMapping.CELL0_UUID:
+                continue
+            output(_('Looking for instances in cell: %s') % cell.identity)
+
+            with context.target_cell(ctxt, cell) as cctxt:
+                self._run_on_instances_batch(
+                    cctxt, cb, instance_uuid=instance_uuid,
+                    expected_attrs=expected_attrs, max_count=max_count,
+                    filters=filters, dry_run=dry_run, output=output)
+
+    def _run_on_instances_batch(self, ctxt, cb, instance_uuid=None,
+                                expected_attrs=None, max_count=50,
+                                filters=None, dry_run=False, output=print):
+        if instance_uuid:
+            max_count = 1
+            unlimited = False
+        elif max_count is not None:
+            try:
+                max_count = int(max_count)
+            except ValueError:
+                max_count = -1
+            unlimited = False
+            if max_count < 1:
+                print(_('Must supply a positive integer for --max-count.'))
+                return 127
+        else:
+            max_count = 50
+            unlimited = True
+            output(_('Running batches of %i until complete') % max_count)
+
+        instance_filters = {'deleted': False}
+        if instance_uuid:
+            instance_filters['uuid'] = instance_uuid
+        if filters:
+            instance_filters.update(filters)
+
+        instances = objects.InstanceList.get_by_filters(
+            ctxt, filters=instance_filters, sort_key='created_at',
+            sort_dir='asc', limit=max_count, expected_attrs=expected_attrs)
+
+        num_processed = 0
+
+        while instances:
+            output(_('Found %s candidate instances.') % len(instances))
+
+            for instance in instances:
+                if cb(ctxt, instance, dry_run=dry_run):
+                    num_processed += 1
+
+            # Make sure we don't go over the max count.
+            if (not unlimited and num_processed == max_count) or instance_uuid:
+                return num_processed
+
+            # Use a marker to get the next page of instances in this cell.
+            marker = instances[len(instances) - 1].uuid
```

Comment:
Why `instances[len(instances) - 1]` and not just `instances[-1]`? Does `collections.Sequence`, which `InstanceList` seems to be based on, not support this?